sir_cake/store/views.py

```diff
@@ -221,3 +221,179 @@ def _get_preorder_or_none(request):
+    }
+
+
+def _get_user_address_instance_or_none(request):
+    # Address instance is None before user makes the first successful payment
+    if request.user.is_anonymous:
+        try:
+            return ShippingAddress.objects.get(
+                user_anon=anonymous_user_with_saved_session(request))
+        except ObjectDoesNotExist:
+            return None
+    try:
+        return ShippingAddress.objects.get(user=request.user)
+    except ObjectDoesNotExist:
+        return None
+
+
+def _create_shipping_address_form(request, address_instance):
+    if request.method == 'POST':
+        if address_instance is None:
+            return ShippingAddressForm(request.POST)
+        return ShippingAddressForm(request.POST, instance=address_instance)
+
+    else:
+        if address_instance is None:
+            return ShippingAddressForm()
+        return ShippingAddressForm(instance=address_instance)
+
+
+def _create_name_form(request):
+    if request.method == 'POST':
+        if request.user.is_anonymous:
+            user = anonymous_user_with_saved_session(request)
+            return AnonymousUserUpdateNamesForm(request.POST, instance=user)
+        return UserUpdateNamesForm(request.POST, instance=request.user)
+    else:
+        if request.user.is_anonymous:
+            user = anonymous_user_with_saved_session(request)
+            return AnonymousUserUpdateNamesForm(instance=user)
+        return UserUpdateNamesForm(instance=request.user)
+
+
+def _save_address_data(request, address_form, first_time_saving):
+    if not address_form.is_valid():
+        messages.error(request, "Invalid form data.")
+        return False
+
+    if first_time_saving:
+        shipping_data = address_form.save(commit=False)
+        if request.user.is_anonymous:
+            shipping_data.user_anon = anonymous_user_with_saved_session(
+                request)
+        else:
+            shipping_data.user = request.user
+        shipping_data.save()
+    else:
+        address_form.save()
+    return True
+
+
+def _save_name_data(request, name_form):
+    if not name_form.is_valid():
+        messages.error(request, "Invalid form data.")
+        return False
+    name_form.save()
+    return True
+
+
+def payment(request):
+    # Dummy view -> doesn't handle real payments
+    if anonymous_user_without_session(request):
+        return redirect('choose_purchasing_mode')
+
+    context = _context_my_bag_total(request)
+    if context is None:
+        return render(request, 'store/payment.html', {'no_items': True})
+
+    if _shipping_data_is_missing(request) == True:
+        messages.error(request, 'Please, enter shipping data')
+        return redirect('shipping')
+
+    preorder = _get_preorder_or_none(request)
+    context['step'] = 3
+    context['form'] = PaymentForm()
+
+    # TODO cron job
+    # When basket expires delete basket and order
+    # _delete_order_and_basket_items(preorder)
+
+    if request.method == "POST":
```

Comment:
I think it would be a bit better to handle `if request.method == "POST":` as soon as possible to avoid having to do duplicate work in POST requests when the form is valid.